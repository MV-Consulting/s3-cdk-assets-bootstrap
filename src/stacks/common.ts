import {
  aws_iam as iam,
} from 'aws-cdk-lib';

export interface UploadRolesConfig {
  uploadRole: iam.Role;
  assetObjectKeysPattern: string;
}

export const assetBucketRegions = [
  'us-east-1', // US East (N. Virginia),
  'eu-west-1', // Europe (Ireland),
  'us-west-2', // US West (Oregon),
  // new regions since 2023-09 for SES
  'eu-central-1', // Europe (Frankfurt),
  // 'eu-west-2', // Europe (London),
  'us-east-2', // US East (Ohio),
  'ca-central-1', // Canada (Central),
  'ap-northeast-1', // Asia Pacific (Tokyo),
  'ap-southeast-1', // Asia Pacific (Singapore),
  'ap-southeast-2', // Asia Pacific (Sydney),
];