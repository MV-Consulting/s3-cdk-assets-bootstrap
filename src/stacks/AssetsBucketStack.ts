import {
  Stack,
  StackProps,
  aws_s3 as s3,
  aws_iam as iam,
  Duration,
  RemovalPolicy,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { UploadRolesConfig } from './common';

export interface AssetsBucketStackProps extends StackProps {
  stage: string;
  devRetentionDays?: number;
  region: string;
  prefix: string;
  uploadRoleConfigs: UploadRolesConfig[];
}

export class AssetsBucketStack extends Stack {
  constructor(scope: Construct, id: string, props: AssetsBucketStackProps) {
    super(scope, id, props);

    const stage = props.stage;
    const assetBucketRegion = props.region;
    const assetBucketPrefix = props.prefix;
    const uploadRoleConfigs = props.uploadRoleConfigs;

    const lifecycleRules: s3.LifecycleRule[] = [];
    if (stage === 'dev') {
      const devRetentionDays = props.devRetentionDays || 14;
      lifecycleRules.push({
        expiration: Duration.days(devRetentionDays),
      });
    }

    const assetBucket = new s3.Bucket(this, `${assetBucketPrefix}-${assetBucketRegion}`, {
      bucketName: `${assetBucketPrefix}-${assetBucketRegion}`,
      removalPolicy: stage === 'dev' ? RemovalPolicy.DESTROY : RemovalPolicy.RETAIN,
      lifecycleRules: lifecycleRules,
      versioned: false,
      blockPublicAccess: {
        blockPublicAcls: false,
        blockPublicPolicy: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
      },
      publicReadAccess: true,
    });

    for (const uploadRoleConfig of uploadRoleConfigs) {
      const uploadRole = uploadRoleConfig.uploadRole;
      const assetObjectKeysPattern = uploadRoleConfig.assetObjectKeysPattern;
      assetBucket.grantPut(uploadRole, assetObjectKeysPattern);

      uploadRole.addToPolicy(new iam.PolicyStatement({
        actions: [
          's3:GetBucketLocation',
          's3:ListBucket',
        ],
        resources: [assetBucket.bucketArn],
      }));
    }

  }
}