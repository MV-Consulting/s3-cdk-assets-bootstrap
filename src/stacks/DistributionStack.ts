import { GithubActionsIdentityProvider, GithubActionsRole } from 'aws-cdk-github-oidc';
import {
  CfnOutput,
  Stack,
  StackProps,
  aws_s3 as s3,
  aws_iam as iam,
  Duration,
  RemovalPolicy,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { UploadRolesConfig } from './common';

export interface GithubRepoProps {
  owner: string;
  repo: string;
  filter: string;
  assetObjectKeysPattern: string;
  releaseObjectKeysPattern: string;
}

export interface DistributionStackProps extends StackProps {
  stage: string;
  devRetentionDays?: number;
  githubRepoProps: GithubRepoProps[];
  releaseBucketProps: {
    prefix: string;
  };
}

export class DistributionStack extends Stack {
  public readonly uploadRoleConfigs: UploadRolesConfig[];

  constructor(scope: Construct, id: string, props: DistributionStackProps) {
    super(scope, id, props);

    const stage = props.stage;
    const bucketPrefix = props.releaseBucketProps.prefix;

    // 1. create role
    const provider = GithubActionsIdentityProvider.fromAccount(
      this,
      'GithubProvider',
    );

    const lifecycleRules: s3.LifecycleRule[] = [];
    if (stage === 'dev') {
      const devRetentionDays = props.devRetentionDays || 14;
      lifecycleRules.push({
        expiration: Duration.days(devRetentionDays),
      });
    }

    // 2. create release bucket
    // release-bucket
    const releaseBucket = new s3.Bucket(this, 'release-bucket', {
      bucketName: `${bucketPrefix}-releases`,
      versioned: false,
      removalPolicy: stage === 'dev' ? RemovalPolicy.DESTROY : RemovalPolicy.RETAIN,
      lifecycleRules: lifecycleRules,
      blockPublicAccess: {
        blockPublicAcls: false,
        blockPublicPolicy: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
      },
      publicReadAccess: true,
    });

    // 3. grant permissions
    const uploadRoleConfigs: UploadRolesConfig[] = [];
    for (const githubRepoProps of props.githubRepoProps) {
      const { owner, repo, filter, assetObjectKeysPattern, releaseObjectKeysPattern } = githubRepoProps;
      const roleName = `ghActionOicd-${stage}-${owner}-${repo}`;
      const uploadRole = new GithubActionsRole(this, roleName, {
        maxSessionDuration: Duration.hours(1),
        provider: provider,
        owner: owner,
        repo: repo,
        filter: filter,
      });

      releaseBucket.grantPut(uploadRole, releaseObjectKeysPattern);
      // see https://github.com/aws/aws-cdk/issues/6808#issuecomment-675198204
      uploadRole.addToPolicy(new iam.PolicyStatement({
        actions: [
          's3:GetBucketLocation',
          's3:ListBucket',
        ],
        resources: [releaseBucket.bucketArn],
      }));

      new CfnOutput(this, `${roleName}-output`, {
        value: uploadRole.roleArn,
      });

      uploadRoleConfigs.push({
        uploadRole: uploadRole,
        assetObjectKeysPattern: assetObjectKeysPattern,
      });
    }
    this.uploadRoleConfigs = uploadRoleConfigs;
  };
}