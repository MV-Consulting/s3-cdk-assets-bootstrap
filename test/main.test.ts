import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DistributionStack } from '../src/stacks/DistributionStack';

test('Snapshot', () => {
  const app = new App();

  const distributionStack = new DistributionStack(app, 'test', {
    stage: 'dev',
    githubRepoProps: [
      {
        owner: 'owner',
        repo: 'repo',
        filter: 'filter',
        assetObjectKeysPattern: 'assetObjectKeysPattern',
        releaseObjectKeysPattern: 'releaseObjectKeysPattern',
      },
    ],
    releaseBucketProps: {
      prefix: 'prefix',
    },
  });

  const templateDistribution = Template.fromStack(distributionStack);
  expect(templateDistribution.toJSON()).toMatchSnapshot();
});