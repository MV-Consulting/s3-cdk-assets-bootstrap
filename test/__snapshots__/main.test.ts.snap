// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Outputs": {
    "ghActionOicddevownerrepooutput": {
      "Value": {
        "Fn::GetAtt": [
          "ghActionOicddevownerrepo53171B57",
          "Arn",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ghActionOicddevownerrepo53171B57": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                },
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": "repo:owner/repo:filter",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":oidc-provider/token.actions.githubusercontent.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "MaxSessionDuration": 3600,
      },
      "Type": "AWS::IAM::Role",
    },
    "ghActionOicddevownerrepoDefaultPolicyB8F181FF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "releasebucket26BB7298",
                        "Arn",
                      ],
                    },
                    "/releaseObjectKeysPattern",
                  ],
                ],
              },
            },
            {
              "Action": [
                "s3:GetBucketLocation",
                "s3:ListBucket",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "releasebucket26BB7298",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ghActionOicddevownerrepoDefaultPolicyB8F181FF",
        "Roles": [
          {
            "Ref": "ghActionOicddevownerrepo53171B57",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "releasebucket26BB7298": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketName": "prefix-releases",
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 14,
              "Status": "Enabled",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": false,
          "BlockPublicPolicy": false,
          "IgnorePublicAcls": false,
          "RestrictPublicBuckets": false,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "releasebucketPolicy3CD656B4": {
      "Properties": {
        "Bucket": {
          "Ref": "releasebucket26BB7298",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "releasebucket26BB7298",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
